File: android/app/src/main/java/com/tdrevans/nexttrain/WidgetSettingsStore.java
package com.tdrevans.nexttrain;

import android.content.Context;
import android.content.SharedPreferences;
import org.json.JSONObject;

public final class WidgetSettingsStore {

  private static final String PREFS = "next_train_widget";
  private static final String KEY_SETTINGS = "settings_json";
  private static final String KEY_SNAPSHOT = "snapshot_json";

  private WidgetSettingsStore() {}

  public static void saveSettings(Context context, String settingsJson) {
    prefs(context).edit().putString(KEY_SETTINGS, settingsJson).apply();
  }

  public static String readSettings(Context context) {
    return prefs(context).getString(KEY_SETTINGS, null);
  }

  public static void saveSnapshot(Context context, JSONObject snapshot) {
    prefs(context).edit().putString(KEY_SNAPSHOT, snapshot.toString()).apply();
  }

  public static JSONObject readSnapshot(Context context) {
    try {
      String raw = prefs(context).getString(KEY_SNAPSHOT, null);
      if (raw == null || raw.isEmpty()) {
        return null;
      }
      return new JSONObject(raw);
    } catch (Exception error) {
      return null;
    }
  }

  public static void saveLastRefreshMs(Context context, long value) {
    prefs(context).edit().putLong(KEY_LAST_REFRESH_MS, value).apply();
  }

  public static long readLastRefreshMs(Context context) {
    return prefs(context).getLong(KEY_LAST_REFRESH_MS, 0L);
  }

  private static final String KEY_LAST_REFRESH_MS = "last_refresh_ms";

  private static SharedPreferences prefs(Context context) {
    return context.getApplicationContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
  }
}

File: ios/Shared/WidgetSettingsStore.swift
import Foundation

enum WidgetSettingsStore {
    private static let settingsKey = "settings_json"
    private static let snapshotKey = "snapshot_json"
    private static let lastRefreshKey = "last_refresh_ms"

    private static var defaults: UserDefaults? {
        UserDefaults(suiteName: WidgetAppGroup.identifier)
    }

    static func saveSettings(_ settingsJson: String) {
        defaults?.set(settingsJson, forKey: settingsKey)
    }

    static func readSettings() -> String? {
        defaults?.string(forKey: settingsKey)
    }

    static func saveSnapshot(_ snapshot: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: snapshot),
              let raw = String(data: data, encoding: .utf8) else {
            return
        }
        defaults?.set(raw, forKey: snapshotKey)
    }

    static func readSnapshot() -> [String: Any]? {
        guard let raw = defaults?.string(forKey: snapshotKey),
              let data = raw.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return json
    }

    static func saveLastRefreshMs(_ value: Int64) {
        defaults?.set(value, forKey: lastRefreshKey)
    }

    static func readLastRefreshMs() -> Int64 {
        Int64(defaults?.double(forKey: lastRefreshKey) ?? 0)
    }
}
